src/spaces/homspace.jl

```diff
@@ -134,3 +140,119 @@ function compose(W::HomSpace{S}, V::HomSpace{S}) where {S}
     return HomSpace(codomain(W), domain(V))
 end
+
+# Block and fusion tree ranges: structure information for building tensors
+#--------------------------------------------------------------------------
+struct FusionBlockStructure{I,F₁,F₂}
+    totaldim::Int
+    blockstructure::SectorDict{I,Tuple{Tuple{Int,Int},UnitRange{Int}}}
+    fusiontreelist::Vector{Tuple{F₁,F₂}}
+    fusiontreestructure::Vector{Tuple{Tuple{Int,Int},Tuple{Int,Int},Int}}
+    fusiontreeindices::FusionTreeDict{Tuple{F₁,F₂},Int}
+end
+
+abstract type CacheStyle end
+struct NoCache <: CacheStyle end
+struct TaskLocalCache{D<:AbstractDict} <: CacheStyle end
+struct GlobalLRUCache <: CacheStyle end
+
+function CacheStyle(I::Type{<:Sector})
+    return GlobalLRUCache()
+end
+
+fusionblockstructure(W::HomSpace) = fusionblockstructure(W, CacheStyle(sectortype(W)))
+
+function fusionblockstructure(W::HomSpace, ::NoCache)
+    codom = codomain(W)
+    dom = domain(W)
+    N₁ = length(codom)
+    N₂ = length(dom)
+    I = sectortype(W)
+    F₁ = fusiontreetype(I, N₁)
+    F₂ = fusiontreetype(I, N₂)
+
+    # output structure
+    blockstructure = SectorDict{I,Tuple{Tuple{Int,Int},UnitRange{Int}}}() # size, range
+    fusiontreelist = Vector{Tuple{F₁,F₂}}()
+    fusiontreestructure = Vector{Tuple{Tuple{Int,Int},Tuple{Int,Int},Int}}() # size, strides, offset
+
+    # temporary data structures
+    splittingtrees = Vector{F₁}()
+    splittingstructure = Vector{Tuple{Int,Int}}()
+
+    # main computational routine
+    blockoffset = 0
+    for c in blocksectors(W)
+        empty!(splittingtrees)
+        empty!(splittingstructure)
+
+        offset₁ = 0
+        for f₁ in fusiontrees(codom, c)
+            push!(splittingtrees, f₁)
+            d₁ = dim(codom, f₁.uncoupled)
+            push!(splittingstructure, (offset₁, d₁))
+            offset₁ += d₁
+        end
+        blockdim₁ = offset₁
+        strides = (1, blockdim₁)
+
+        offset₂ = 0
+        for f₂ in fusiontrees(dom, c)
+            s₂ = f₂.uncoupled
+            d₂ = dim(dom, s₂)
+            for (f₁, (offset₁, d₁)) in zip(splittingtrees, splittingstructure)
+                push!(fusiontreelist, (f₁, f₂))
+                totaloffset = blockoffset + offset₂ * blockdim₁ + offset₁
+                push!(fusiontreestructure, ((d₁, d₂), strides, totaloffset))
+            end
+            offset₂ += d₂
+        end
+        blockdim₂ = offset₂
+        blocksize = (blockdim₁, blockdim₂)
+        blocklength = blockdim₁ * blockdim₂
+        blockrange = (blockoffset + 1):(blockoffset + blocklength)
+        blockoffset = last(blockrange)
+        blockstructure[c] = (blocksize, blockrange)
+    end
+
+    fusiontreeindices = sizehint!(FusionTreeDict{Tuple{F₁,F₂},Int}(),
+                                  length(fusiontreelist))
+    for (i, f₁₂) in enumerate(fusiontreelist)
+        fusiontreeindices[f₁₂] = i
+    end
+    totaldim = blockoffset
+    structure = FusionBlockStructure{I,F₁,F₂}(totaldim, blockstructure,
```

Comment:
This structure makes it quite difficult to figure out what stridedview corresponds with [:,f2] (for a given f2.coupled). So what used to be the old rowr/colr takes a bit of compute to get out again. Not really a problem on it's own, but I'm surprised that that operation is never needed (for example if I want to permute two spaces in the codomain, do we then iterate over all possible (f1,f2) pairs?)

Reply:
The `permute` function does indeed not specialise for permutations that do not cross the domain or the codomain. If I remember correctly, the idea, as dicussed with @lkdvos, was that all `(f1,f2)` pairs are now easily associated with a linear index, such that the current caches for permutations (and transpositions) can take the form of a sparse matrix, which would then also make it easier to compute their inverse etc.

I do see what you mean; that doesn't really come up in any of the implementations so far (it's always acting on individual (f1,f2) pairs), but it is probably needed for what you are trying to do. I am happy to further revise the structure.

Reply:
I guess the claim is that you never really need the `rowr` and `colr` anymore anyways, since it does not really make sense to first try and obtain a view of the block before obtaining a view of the subblock. Therefore, there is no real difference between permutations within a block (keeping coupled charges fixed) and permutations that do not have this, since these all index into one contiguous vector anyways. Before, you would have to do something like:
1. Get block for `c` (dictionary lookup)
2. Index into that block with `rowr[c][f1]` and `colr[c][f2]`

While now, it should really just be:
1. Index into the data using the structure

The hope being that we reduce dictionary lookups etc.

Additionally, the structure is sorted/stored in a vector, so we should be able to avoid additional dict lookups within `add_transform!` by replacing the `fusiontreemap` with a simple sparse matrix, using integers instead of hashables (this is not yet implemented though). Of course, we still want the option to have "random access" patterns, so there is also still a `fusiontreeindices` dictionary that maps fusiontrees to the correct index. In this way, I would hope to be able to at least optimize `add_transform!` to do a single cache-lookup for the structure, and then have just vector indexing for extracting the correct data. I might try my hand at that implementation if I find some time tonight.

Reply:
You can often gain a lot of speed if you can vectorize a group of similar operations. Consider that you are manipulating a part of the codomain where all blocks [f1,f2] get mapped to [f1,f2'] (so not a function of f1), then you might be able to simultaneously vectorize a group of operations [:,f2] -> [:,f2'], which should be significantly faster.

To give a concrete - but perhaps not relevant - example, let's say you have a tensormap and you want to project out a single fusiontree in the codomain. With the current datastructure, you'd have to go over every (f1,f2) pair and zero out if f1 matches. In the old tensorkit, you could have written something like data[f1.coupled][rowr[f1],:].=0, which should be significantly faster.

All that aside - having indexmanipulations implemented as sparse matrices is going to be amazing!